src/lib.rs

```diff
@@ -363,15 +378,90 @@ where
+    /// then truncating the result to length `k`.  In particular, items with the same count are
+    /// sorted in *increasing* order of their keys.  Further, if `k` is greater than the length of
+    /// the counter then the returned vector will have length equal to that of the counter, not
+    /// `k`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use counter::Counter;
+    /// let counter: Counter<_> = "abracadabra".chars().collect();
+    /// let top3 = counter.k_most_common_ordered(3);
+    /// assert_eq!(top3, vec![('a', 5), ('b', 2), ('r', 2)]);
+    /// ```
+    ///
+    /// # Time complexity
+    ///
+    /// This method can be much more efficient than [`most_common_ordered`] when *k* is much
+    /// smaller than the length of the counter *n*.  When *k* = 1 the algorithm is equivalent
+    /// to finding the minimum (or maximum) of *n* items, which requires *n* \- 1 comparisons.  For
+    /// a fixed value of *k* > 1, the number of comparisons scales with *n* as *n* \+ *O*(log *n*)
+    /// and the number of swaps scales as *O*(log *n*).  As *k* approaches *n*, this algorithm
+    /// approaches a heapsort of the *n* items, which has complexity *O*(*n* \* log *n*).
+    ///
+    /// [`most_common_ordered`]: Counter::most_common_ordered
+    pub fn k_most_common_ordered(&self, k: usize) -> Vec<(T, N)> {
+        use std::cmp::Reverse;
+
+        if k == 0 {
+            return vec![];
+        }
+
+        // Clone the counts as we iterate over the map to eliminate an extra indirection when
+        // comparing counts.  This will be an improvement in the typical case where `N: Copy`.
+        // Defer cloning the keys until we have selected the top `k` items so that we clone only
+        // `k` keys instead of all of them.
+        let mut items = self.map.iter().map(|(t, n)| (Reverse(n.clone()), t));
+
+        // Step 1. Make a heap out of the first `k` items; this makes O(k) comparisons.
+        let mut heap: BinaryHeap<_> = items.by_ref().take(k).collect();
+
+        // Step 2. Successively compare each of the remaining `n - k` items to the top of the heap,
+        // replacing the root (and subsequently sifting down) whenever the item is less than the
+        // root.  This takes at most n - k + k * (1 + log2(k)) * (H(n) - H(k)) comparisons, where
+        // H(i) is the ith [harmonic number](https://en.wikipedia.org/wiki/Harmonic_number).  For
+        // fixed `k`, this scales as *n* + *O*(log(*n*)).
+        items.for_each(|item| {
+            // If `items` is nonempty at this point then we know the heap contains `k > 0`
+            // elements.
+            let mut root = heap.peek_mut().expect("the heap is empty");
+            if *root > item {
+                *root = item;
+            }
+        });
+
+        // Step 3. Sort the items in the heap with the second phases of heapsort.  The number of
+        // comparisons is 2 * k * log2(k) + O(k).
+        heap.into_sorted_vec()
```

Comment:
Too bad about https://github.com/rust-lang/rust/pull/76234, huh?

Reply:
Yeah, exactly.  That would be useful here.

Incidentally, I orginally used `.into_vec()` followed by `slice::sort_unstable()` instead of `.into_sorted_vec()` as suggested in the discussion on that PR:
```rust
let mut topk = heap.into_vec();
topk.sort_unstable();
```
In my benchmarks, for small heaps `.into_sorted_vec()` is just slightly faster, but as the heap size grows it becomes faster to take the vector from the heap and use `slice::sort_unstable()`.  I was on the fence for a while about what to do here.  In the end I thought it was a little more direct to just use `.into_sorted_vec()`.  Plus, then I realized that as *k* approaches *n* the algorithm becomes a heapsort.  So if nothing else it is a little easier to think about what is happening, at least for me.

Anyway, it would certainly be an option to take the vector from the heap and sort it instead of finishing the heapsort in step 3 if you like that approach better.  But as long as *k* is small compared to *n* it doesn't matter which way we do it, it is much better than sorting the entire counter.